flow: Add support for more stages to `loki.process` (#2711)
grafana/agent: docs/sources/flow/reference/components/loki.process.md

The `stage` block does not support any arguments and is configured only via
inner blocks.

### docker block

The `docker` inner block enables a predefined pipeline which reads log lines in
the standard format of Docker log files.

The `docker` block does not support any arguments or inner blocks, but is
always empty.

```river
stage {
	docker {}
}
```

Docker log entries are formatted as JSON with the following keys:

> 1) `log`: The content of log line
> 2) `stream`: Either `stdout` or `stderr`
> 3) `time`: The timestamp string of the log line

Reviewer: This isn't an ordered list of instructions, so it should be unordered:

```suggestion
* `log`: The content of log line
* `stream`: Either `stdout` or `stderr`
* `time`: The timestamp string of the log line
```

Then, given the following log line, the following key-value pairs would be
created in the shared map of extracted data:

```
{"log":"log message\n","stream":"stderr","time":"2019-04-30T02:12:41.8443515Z"}

output: log message\n
stream: stderr
timestamp: 2019-04-30T02:12:41.8443515
```

### cri block

The `cri` inner block enables a predefined pipeline which reads log lines using
the CRI logging format.

The `cri` block does not support any arguments or inner blocks, but is always
empty.

```river
stage {
	cri {}
}
```

CRI specifies log lines as single space-delimited values with the following
components:

> 1) `time`: The timestamp string of the log
> 2) `stream`: Either `stdout` or `stderr`
> 3) `flags`: CRI flags including `F` or `P`
> 4) `log`: The contents of the log line

Reviewer: ```suggestion
* `time`: The timestamp string of the log
* `stream`: Either `stdout` or `stderr`
* `flags`: CRI flags including `F` or `P`
* `log`: The contents of the log line
```

The `stage` block does not support any arguments and is configured only via
inner blocks.

### docker block

The `docker` inner block enables a predefined pipeline which reads log lines in
the standard format of Docker log files.

> The `docker` block does not support any arguments or inner blocks, but is

Reviewer 2: ```suggestion
The `docker` block does not support any arguments or inner blocks, so it is
```

always empty.

```river
stage {
	docker {}
}
```

Docker log entries are formatted as JSON with the following keys:

* `log`: The content of log line
* `stream`: Either `stdout` or `stderr`
* `time`: The timestamp string of the log line

> Then, given the following log line, the following key-value pairs would be
> created in the shared map of extracted data:

Reviewer 2: ```suggestion
Given the following log line, the subsequent key-value pairs are
created in the shared map of extracted data:
```

```
{"log":"log message\n","stream":"stderr","time":"2019-04-30T02:12:41.8443515Z"}

output: log message\n
stream: stderr
timestamp: 2019-04-30T02:12:41.8443515
```

### cri block

The `cri` inner block enables a predefined pipeline which reads log lines using
the CRI logging format.

The `cri` block does not support any arguments or inner blocks, but is always
empty.

```river
stage {
	cri {}
}
```

CRI specifies log lines as single space-delimited values with the following
components:

* `time`: The timestamp string of the log
* `stream`: Either `stdout` or `stderr`
* `flags`: CRI flags including `F` or `P`
* `log`: The contents of the log line

> Given the following log line, the following key-value pairs would be created in
> the shared map of extracted data:

Reviewer 2: ```suggestion
Given the following log line, the subsequent key-value pairs are created in
the shared map of extracted data:
```

------------- | --------- | ------------------------------------------------------------------- | ------- | --------
`expression`  | `string`  | A valid RE2 regular expression. Each capture group must be named.   |         | yes
`source`      | `string`  | Name from extracted data to parse. If empty, uses the log message.  | `""`    | no

The `expression` field needs to be a Go RE2 regex string. Every capture group (re) is set into the extracted map, so it must be named like: `(?P<name>re)`. The name of the capture group is then used as the key in the extracted map.

<!--
We don't care about YAML, what does River do instead???

Because of how YAML treats backslashes in double-quoted strings, note that all backslashes in a regex expression must be escaped when using double quotes. For example, all of these are valid:

expression: \w*
expression: '\w*'
expression: "\\w*"
But these are not:

expression: \\w* (only escape backslashes when using double quotes)
expression: '\\w*' (only escape backslashes when using double quotes)
expression: "\w*" (backslash must be escaped)
-->

If the `source` is empty, then the stage uses attempts to parse the log line
itself.

> Given the following log line and regex stage, the extracted values are:

Reviewer 2: ```suggestion
Given the following log line and regex stage, the extracted values are shown below:
```

```
2019-01-01T01:00:00.000000001Z stderr P i'm a log message!
stage {
  regex {
    expression = "^(?s)(?P<time>\\S+?) (?P<stream>stdout|stderr) (?P<flags>\\S+?) (?P<content>.*)$"
  }
}

time: 2019-01-01T01:00:00.000000001Z,
stream: stderr,
flags: P,
content: i'm a log message
```

On the other hand, if the `source` value is set, then the regex is applied to
the value stored in the shared map under that name.

Let's see what happens when the following log line is put through this
two-stage pipeline:
```
{"timestamp":"2022-01-01T01:00:00.000000001Z"}

stage {
  json {
    expressions = { time = "timestamp" }
  }
}
stage {
  regex {
    expression = "^(?P<year>\\d+)"
    source     = "time"
  }
}
```

> The first stage would add the following key-value pairs into the extracted map:

Reviewer 2: ```suggestion
The first stage adds the following key-value pairs into the extracted map:
```

```
time: 2022-01-01T01:00:00.000000001Z
```

> Then, the regex stage would parse the value for time from the shared values and
> append the following key-value pairs back into the extracted values map:

Reviewer 2: ```suggestion
Then, the regex stage parses the value for time from the shared values and
appends the subsequent key-value pairs back into the extracted values map:
```

Author: Yes, fixed it for the previous instance as well!

`format`            | `string`       | Determines how to parse the source string.                  |           | yes
`fallback_formats`  | `list(string)` | Fallback formats to try if the `format` field fails.        | `[]`      | no
`location`          | `string`       | IANA Timezone Database location to use when parsing.        | `""`      | no
`action_on_failure` | `string`       | What to do when the timestamp can't be extracted or parsed. | `"fudge"` | no

The `source` field defines which value from the shared map of extracted values
the stage should attempt to parse as a timestamp.

The `format` field defines _how_ that source should be parsed.

First off, the `format` can be set to one of the following shorthand values for
commonly-used forms:
```
ANSIC: Mon Jan _2 15:04:05 2006
UnixDate: Mon Jan _2 15:04:05 MST 2006
RubyDate: Mon Jan 02 15:04:05 -0700 2006
RFC822: 02 Jan 06 15:04 MST
RFC822Z: 02 Jan 06 15:04 -0700
RFC850: Monday, 02-Jan-06 15:04:05 MST
RFC1123: Mon, 02 Jan 2006 15:04:05 MST
RFC1123Z: Mon, 02 Jan 2006 15:04:05 -0700
RFC3339: 2006-01-02T15:04:05-07:00
RFC3339Nano: 2006-01-02T15:04:05.999999999-07:00

> 

Reviewer 2: ```suggestion
```

```

> Additionally, support for common Unix timestamps is supported with the following format values:

Reviewer 2: The "Additionally" line was showing up as code.

Author: Yes, split it up into different code blocks.

> Unix: 1562708916 or with fractions 1562708916.000000123

Reviewer 2: ```suggestion
```
Unix: 1562708916 or with fractions 1562708916.000000123
```

UnixMs: 1562708916414
UnixUs: 1562708916414123
UnixNs: 1562708916000000123
```

> Otherwise, the field accepts a custom format string that defines how an
> arbitrary reference point in history (Mon Jan 2 15:04:05 -0700 MST 2006) should
> be interpreted by the stage.

Reviewer 2: ```suggestion
Otherwise, the field accepts a custom format string that defines how an
arbitrary reference point in history (for example, `Mon Jan 2 15:04:05 -0700 MST 2006`) should
be interpreted by the stage.
```

Author: This is not an example, should I rephrase it somehow?

The Go language uses this set point time in history `Mon Jan 2 15:04:05 -0700 MST 2006`, as a reference point when parsing time formats.

So if I wanted to format my dates as "YYYY-MM-DD" I should use the reference time to denote that i.e. "2006-01-02". Similarly, if I wanted to format it as "HH:MM", I'd use "03:04 pm".

Reviewer 2: Wow, weird. Should you say something like: 

```suggestion
Otherwise, the field accepts a custom format string that defines how an
arbitrary reference point in history should
be interpreted by the stage. The arbitrary reference point is Mon Jan 2 15:04:05 -0700 MST 2006.
```

The string value of the field is passed directly to the layout parameter in
Go's [`time.Parse`](https://pkg.go.dev/time#Parse) function.

If the custom format has no year component the stage uses the current year,
according to the system's clock.

> The following table shows the supported reference values which should be used
> when defining a custom format.

Reviewer 2: ```suggestion
The following table shows the supported reference values to use
when defining a custom format.
```

Timestamp Component | Format value 
------------------- | --------------
Year                | 06, 2006
Month               | 1, 01, Jan, January
Day                 | 2, 02, _2 (two digits right justified)
Day of the week     | Mon, Monday
Hour                | 3 (12-hour), 03 (12-hour zero prefixed), 15 (24-hour)
Minute              | 4, 04
Second              | 5, 05
Fraction of second  | .000 (ms zero prefixed), .000000 (μs), .000000000 (ns), .999 (ms without trailing zeroes), .999999 (μs), .999999999 (ns)
12-hour period      | pm, PM
Timezone name       | MST
Timezone offset     | -0700, -070000 (with seconds), -07, 07:00, -07:00:00 (with seconds)
Timezone ISO-8601   | Z0700 (Z for UTC or time offset), Z070000, Z07, Z07:00, Z07:00:00

> The `fallback_formats` field allows to define one or more format fields to try
> and parse the timestamp with, if parsing with `format` fails.

Reviewer 2: ```suggestion
The `fallback_formats` field defines one or more format fields to try
and parse the timestamp with, if parsing with `format` fails.
```

The `location` field must be a valid IANA Timezone Database location and
determines in which timezone the timestamp value is interpreted to be in.

The `action_on_failure` field defines what should happen when the source field
doesn't exist in the shared extracted map, or if the timestamp parsing fails.

The supported actions are:

> * fudge (default): change the timestamp to the last known timestamp, summing up 1 nanosecond (to guarantee log entries ordering)
> * skip: do not change the timestamp and keep the time when the log entry has been scraped

Reviewer 2: ```suggestion
* fudge (default): Change the timestamp to the last known timestamp, summing up 1 nanosecond (to guarantee log entries ordering).
* skip: Do not change the timestamp and keep the time when the log entry was scraped.
```

> The following stage fetches the `time` value from the shared values map, parses
> it as a RFC3339 format and set is as the log entry's timestamp.

Reviewer 2: ```suggestion
The following stage fetches the `time` value from the shared values map, parses
it as a RFC3339 format, and sets it as the log entry's timestamp.
```

```
stage {
	timestamp {
		source = "time"
		format = "RFC3339"
	}
}
```

### output block

The `output` inner block configures a processing stage that reads from the
extracted map and changes the content of the log entry that is forwarded
to the next component.

The following arguments are supported:

Name                | Type           | Description                                           | Default   | Required
------------------- | -------------- | ----------------------------------------------------- | --------- | --------
`source`            | `string`       | Name from extracted data to use for the log entry.    |           | yes

> Let's see how this would work for the following log line and three-stage
> pipeline

Reviewer 2: ```suggestion
Let's see how this works for the following log line and three-stage
pipeline:
```
